service/storage_service.cc

```diff
@@ -1974,12 +1974,15 @@ future<> storage_service::join_token_ring(sharded<db::system_distributed_keyspac
     }
 
     slogger.debug("Setting tokens to {}", bootstrap_tokens);
-    co_await mutate_token_metadata([this, &bootstrap_tokens] (mutable_token_metadata_ptr tmptr) -> future<> {
+    co_await mutate_token_metadata([this, &bootstrap_tokens, &replaced_host_id] (mutable_token_metadata_ptr tmptr) -> future<> {
         // This node must know about its chosen tokens before other nodes do
         // since they may start sending writes to this node after it gossips status = NORMAL.
         // Therefore, in case we haven't updated _token_metadata with our tokens yet, do it now.
         tmptr->update_topology(tmptr->get_my_id(), _snitch.local()->get_location(), locator::node::state::normal);
         co_await tmptr->update_normal_tokens(bootstrap_tokens, tmptr->get_my_id());
+        if (replaced_host_id) {
+            tmptr->remove_endpoint(*replaced_host_id);
+        }
     });
```

Comment:
I still have a small problem with the commit message:
> We want to
rely on the invariant that token owners are always in the normal and
leaving states. However, this invariant is broken by a replacing node
because it does not remove the replaced node from topology. Hence,
after joining, the replacing node has topology with a node that is
not a token owner but is in a leaving state (`being_replaced`). We
fix it to prevent the following patch from introducing a regression.

if the invariant is "token owner => normal or leaving"
then a node which is "not token owner ^ leaving" (the being_replaced node) does not break the invariant (if the antecedent is false, as in this case, the implication is trivially satisfied).

Reply:
Right, a different invariant is broken here: "token owner at one time => (normal or leaving => token owner)". This invariant is also broken by a node being marked as normal in `storage_service::on_alive` too early. I'll fix both commit messages.